File: 2015/python/18.py
# Advent of Code 2015, day 18
# (c) aichingert

def sur(p, g):
    ans = 0
    for i in range(-1,2):
        for j in range(-1,2):
            if i == 0 == j:
                continue
            if not (p[0]+i < 0 or p[0]+i >= len(g) or p[1]+j < 0 or p[1]+j >= len(g[p[0]])) and g[p[0]+i][p[1]+j] == '#':
                ans += 1
    return ans

def s_c(g):
    g[0][0] = '#'
    g[0][len(grid[0])-1] = '#'
    g[len(g)-1][0] = '#'
    g[len(g)-1][len(g[0])-1] = '#'
    return g

def solve(g, part):
    for _ in range(100):
        n = [['.' for _ in range(len(g[0]))] for _ in range(len(g))]
        if part: s_c(n)

        ans = 0
        for i in range(len(g)):
            for j in range(len(g[i])):
                if part and (i,j) in [(0,0), (0,len(g[i])-1), (len(g)-1,0), (len(g)-1,len(g[i])-1)]:
                        continue
                nb = sur([i,j], g)
                if g[i][j] == '#':
                    if nb == 2 or nb == 3:
                        n[i][j] = '#'
                    else:
                        n[i][j] = '.'
                else:
                    if g[i][j] == '.' and nb == 3:
                        n[i][j] = '#'
                    else:
                        n[i][j] = '.'
        g = n
    for a in g:
        for e in a:
            if e == '#': ans += 1
    return ans

def get():
    grid = []
    for l in open("../input/18").read().split('\n'):
        grid.append([c for c in l])
    return grid

grid = get()
print("Part 1:",solve(grid,False))
grid = s_c(get())
print("Part 2:",solve(grid,True))

File: 2015/rust/18.rs
// Advent of Code 2015, day 18
// (c) aichingert

fn neighbours(m: &Vec<Vec<char>>, i: usize, j: usize) -> u32 {
    let mut ans = 0;
    for y in -1..2 {
        for x in -1..2 {
            if y == 0 && x == 0 { continue; }
            if (y+(i as i32))>=0 && (x+(j as i32))>=0 && ((y+(i as i32)) as usize)<m.len() && ((x+(j as i32)) as usize)<m[i].len() && m[((y+(i as i32)) as usize)][((x+(j as i32)) as usize)] == '#' {
                ans += 1;
            }
        }
    }
    ans
}

fn corners(m: &mut Vec<Vec<char>>) {
    let l = m.len()-1;
    let lz = m[0].len()-1;
    m[0][0] = '#';
    m[0][lz] = '#';
    m[l][0] = '#';
    m[l][lz] = '#';
}

fn solve(inp: &Vec<Vec<char>>, part: bool) -> usize {
    let mut cur = inp.clone();
    if part { corners(&mut cur); }

    for _ in 0..100 {
        let mut n = cur.clone();
        for i in 0..inp.len() {
            for j in 0..inp[i].len() {
                let arnd = neighbours(&cur, i, j);

                n[i][j] = match cur[i][j] {
                    '#' => if arnd == 2 || arnd == 3 { '#' } else { '.' },
                    _ => if arnd == 3 { '#' } else { '.' },
                };
            }
        }
        cur = n;
        if part { corners(&mut cur); }
    }

    cur.iter().map(|l| l.iter().filter(|ch| *ch == &'#').count()).sum::<usize>()
}

fn main() {
    let inp = std::fs::read_to_string("../input/18").unwrap();
    let inp = inp.lines().map(|l| l.chars().collect::<Vec<char>>()).collect::<Vec<Vec<char>>>();

    println!("Part 1: {}", solve(&inp, false));
    println!("Part 2: {}", solve(&inp, true));
}
